docs/architecture/adr-070-unordered-account.md

```diff
@@ -2,214 +2,137 @@
-i.e. before the current block time. If so, we reject it. Upon inclusion in a block
-during `DeliverTx`, the transaction's hash is set in the map along with it's TTL
-value.
-
-This map is evaluated at the end of each block, e.g. ABCI `Commit`, and all stale
-transactions, i.e. transactions's TTL value who's now beyond the committed block,
-are purged from the map.
-
-An important point to note is that in theory, it may be possible to submit an unordered
-transaction twice, or multiple times, before the transaction is included in a block.
-However, we'll note a few important layers of protection and mitigation:
-
-* Assuming CometBFT is used as the underlying consensus engine and a non-noop mempool
-  is used, CometBFT will reject the duplicate for you.
-* For applications that leverage ABCI++, `ProcessProposal` should evaluate and reject
-  malicious proposals with duplicate transactions.
-* For applications that leverage their own application mempool, their mempool should
-  reject the duplicate for you.
-* Finally, worst case if the duplicate transaction is somehow selected for a block
-  proposal, 2nd and all further attempts to evaluate it, will fail during `DeliverTx`,
-  so worst case you just end up filling up block space with a duplicate transaction.
-
-```golang
-type TxHash [32]byte
+We facilitate replay protection by storing the unordered sequence, in the Cosmos SDK KV store. Upon transaction ingress, we check if the transaction's unordered
+sequence exists in state, or if the TTL value is stale, i.e. before the current block time. If so, we reject it. Otherwise,
+we add the unordered sequence to state. This section of the state will belong to the `x/auth` module.
 
-const PurgeLoopSleepMS = 500
+The state is evaluated during x/auth's `PreBlocker`. All transactions with an unordered sequence earlier than the current block time
+will be deleted.
 
-// UnorderedTxManager contains the tx hash dictionary for duplicates checking,
-// and expire them when block production progresses.
-type UnorderedTxManager struct {
-  // blockCh defines a channel to receive newly committed block time
-  blockCh chan time.Time
-
-  mu sync.RWMutex
-	// txHashes defines a map from tx hash -> TTL value, which is used for duplicate
-	// checking and replay protection, as well as purging the map when the TTL is
-	// expired.
-	txHashes map[TxHash]time.Time
-}
-
-func NewUnorderedTxManager() *UnorderedTxManager {
-  m := &UnorderedTxManager{
-		blockCh:  make(chan time.Time, 16),
-		txHashes: make(map[TxHash]time.Time),
-  }
-
- return m
-}
-
-func (m *UnorderedTxManager) Start() {
-  go m.purgeLoop()
-}
-
-func (m *UnorderedTxManager) Close() error {
-  close(m.blockCh)
-  m.blockCh = nil
-  return nil
+```go
+func (am AppModule) PreBlock(ctx context.Context) (appmodule.ResponsePreBlock, error) {
+	err := am.accountKeeper.GetUnorderedTxManager().RemoveExpired(sdk.UnwrapSDKContext(ctx))
```

Comment:
we should have NPE protection on this for when users do not use this

Reply:
the account keeper currently always has this instantiated, so it wouldn't NPE

you can specify to enable the unordered ante in the ante options, but we'd likely need another option if we want to enable/disable this. might be awkward UX to specify this in two different places though. probably best to walk through this on a call

Reply:
Yeah I see your point - lets just be aware of it when we do code reviews for this.  I think you're right about how it is awkward if users have to set in multiple places

Reply:
Any thoughts here? Looks like this was marked as resolved